ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala

```diff
@@ -91,14 +95,30 @@ private[backend] object H2StorageBackend
       key: String,
       submittedAt: Instant,
       deduplicateUntil: Instant,
-  )(connection: Connection): Int =
-    SQL(SQL_INSERT_COMMAND)
-      .on(
-        "deduplicationKey" -> key,
-        "submittedAt" -> Timestamp.instantToMicros(submittedAt),
-        "deduplicateUntil" -> Timestamp.instantToMicros(deduplicateUntil),
-      )
-      .executeUpdate()(connection)
+  )(connection: Connection)(implicit loggingContext: LoggingContext): Int = {
+
+    // Under the default READ_COMMITTED isolation level used for the indexdb, when a deduplication
+    // upsert is performed simultaneously from multiple threads, the query fails with
+    // JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation
+    // Simple retry helps
+    def retry[T](op: => T): T =
```

Comment:
This mechanism should be moved to a common util to avoid duplication.

Reply:
This is the PR for porting a fix from daml-main to the release candidate. There is probably something to be said for keeping the daml-main and the port fixes consistent, or?

Reply:
🤦 
You're right, sorry!

Reply:
I disagree with moving it out. I wanted to achieve the opposite effect: have this change function-local so that it can be easily removed when a comprehensive retry is introduced across the board.